code/lib/test/src/spy.ts

```diff
@@ -1,17 +1,58 @@
+/* eslint-disable @typescript-eslint/no-shadow */
+import type { MockInstance } from '@vitest/spy';
 import {
-  spyOn,
+  spyOn as vitestSpyOn,
   isMockFunction,
-  fn,
+  fn as vitestFn,
   mocks,
   type MaybeMocked,
   type MaybeMockedDeep,
   type MaybePartiallyMocked,
   type MaybePartiallyMockedDeep,
 } from '@vitest/spy';
+import type { SpyInternalImpl } from 'tinyspy';
+import * as tinyspy from 'tinyspy';
 
 export type * from '@vitest/spy';
 
-export { spyOn, isMockFunction, fn, mocks };
+export { isMockFunction, mocks };
+
+type Listener = (mock: MockInstance, args: unknown[]) => void;
+const listeners = new Set<Listener>();
+
+export function onMockCall(callback: Listener): () => void {
+  listeners.add(callback);
+  return () => void listeners.delete(callback);
+}
+
+// @ts-expect-error Make sure we export the exact same type as @vitest/spy
+export const spyOn: typeof vitestSpyOn = (...args) => {
+  const mock = vitestSpyOn(...(args as Parameters<typeof vitestSpyOn>));
+  return reactiveMock(mock);
+};
+
+// @ts-expect-error Make sure we export the exact same type as @vitest/spy
+export const fn: typeof vitestFn = (implementation) => {
+  const mock = implementation ? vitestFn(implementation) : vitestFn();
+  return reactiveMock(mock);
+};
+
+function reactiveMock(mock: MockInstance) {
+  const reactive = listenWhenCalled(mock);
+  const originalMockImplementation = reactive.mockImplementation.bind(null);
```

Comment:
cc @kasperpeulen

Reply:
We added a fix for this 8.2, could you check that @dmy-leanix?

Reply:
@kasperpeulen Currently can't check in my repo as in 8.2 some paths were changed and during running storybook I have:

```
Cannot find module 'storybook/internal/common'
```

Looks like related to this: https://github.com/storybookjs/storybook/commit/3a1e61cefec8dfd350d42db5598b2d0d9159a878

Most probably nx or angular has own usages and this change wasn't backward compatible to it. And I can't update nx and angular right now

Reply:
@dmy-leanix I'd need a repro to investigate.

I suspect you may have multiple versions of the `storybook` package.

Reply:
@dmy-leanix can you run “npx storybook doctor” and share the output?

Reply:
@yannbf  Doctor command:
```
$ npx storybook@latest doctor --config-dir .storybook
🩺 The doctor is checking the health of your Storybook..

🥳 Your Storybook project looks good!
You can always recheck the health of your project by running:
npx storybook doctor
```

@ndelangen checked in package-lock.json, it's one version:
```
$ npm list @storybook/angular
[...project...]
└── @storybook/angular@8.2.2
```

Full stack trace:
```
[0]  >  NX   Cannot find module 'storybook/internal/common'
[0] 
[0]    Require stack:
[0]    - .../project/node_modules/@storybook/angular/dist/builders/start-storybook/index.js
[0]    - .../project/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js
[0]    - .../project/node_modules/@angular-devkit/architect/node/index.js
[0]    - .../project/node_modules/nx/src/adapter/ngcli-adapter.js
[0]    - .../project/node_modules/nx/src/command-line/run/run.js
[0]    - .../project/node_modules/nx/bin/run-executor.js
```

With --verbose flag:
```
Error: Cannot find module 'storybook/internal/common'
Require stack:
- .../project/node_modules/@storybook/angular/dist/builders/start-storybook/index.js
- .../project/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js
- .../project/node_modules/@angular-devkit/architect/node/index.js
- .../project/node_modules/nx/src/adapter/ngcli-adapter.js
- .../project/node_modules/nx/src/command-line/run/run.js
- .../project/node_modules/nx/bin/run-executor.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at Module.require (.../project/node_modules/nx/src/adapter/compat.js:61:40)
    at require (node:internal/modules/cjs/helpers:121:18)
    at Object.<anonymous> (.../project/node_modules/@storybook/angular/dist/builders/start-storybook/index.js:8:18)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Module._load (node:internal/modules/cjs/loader:960:12)
```

Reply:
Can you show `npm list storybook` and `npm why storybook`?

Reply:
@kasperpeulen "No dependencies found matching storybook", do I need `storybook` additionally to `@storybook/*`?

Reply:
@kasperpeulen yes, after adding `storybook` dependency it works fine, didn't know that this dependency should also be add

Reply:
Storybook for angular (as well as a lot of other packages) has a peerDependency on `storybook` starting `8.2`.
https://github.com/storybookjs/storybook/blob/3b04c5b4509f48824591a9814b0a82b0b184336d/code/frameworks/angular/package.json#L92

The `storybook` package has been a required devDependency since `7.0`.

It should have been added by this auto-migration:
https://github.com/storybookjs/storybook/blob/e35b884b9ef3811253ff4d3d096dc97db66cc001/code/lib/cli/src/automigrate/fixes/sb-binary.ts#L23-L107
...when you upgraded from 6.x to 7.x.

Reply:
@ndelangen  thank you! yes, strange I didn't get error related to this before

Reply:
@ndelangen This automigration didn't run for Nx users. @yannbf fixed that here:
https://github.com/storybookjs/storybook/pull/28601